shared/src/sys/mod.rs

```diff
@@ -38,3 +99,129 @@ impl<'a> TryFrom<&'a crate::econ::TokenAmount> for TokenAmount {
         })
     }
 }
+
+impl fmt::Display for TokenAmount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", crate::econ::TokenAmount::from(self))
+    }
+}
+
+impl Add<Self> for TokenAmount {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        crate::econ::TokenAmount::from(self)
+            .add(&crate::econ::TokenAmount::from(rhs))
+            .try_into()
+            .unwrap()
+    }
+}
+
+impl Sub<Self> for TokenAmount {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        // panic if rhs > self?
+        crate::econ::TokenAmount::from(self)
+            .sub(&crate::econ::TokenAmount::from(rhs))
+            .try_into()
+            .unwrap()
+    }
+}
+
+impl Zero for TokenAmount {
+    #[inline]
+    fn zero() -> TokenAmount {
+        TokenAmount { hi: 0, lo: 0 }
+    }
+
+    #[inline]
+    fn set_zero(&mut self) {
+        self.hi = 0;
+        self.lo = 0;
+    }
+
+    #[inline]
+    fn is_zero(&self) -> bool {
+        self.hi == 0 && self.lo == 0
+    }
+}
+
+impl Default for TokenAmount {
+    #[inline]
+    fn default() -> TokenAmount {
+        Zero::zero()
+    }
+}
+
+impl PartialEq<Self> for TokenAmount {
+    fn eq(&self, other: &Self) -> bool {
+        self.hi == other.hi && self.lo == other.lo
+    }
+}
+
+impl Eq for TokenAmount {}
+
+impl Hash for TokenAmount {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.lo.hash(state);
+        if self.hi != 0 {
+            self.hi.hash(state);
+        }
+    }
+}
+
+impl Mul<u64> for TokenAmount {
+    type Output = TokenAmount;
+
+    fn mul(self, rhs: u64) -> Self::Output {
+        let eta: crate::econ::TokenAmount = self.into();
+        crate::econ::TokenAmount::from(eta * rhs)
+            .try_into()
+            .unwrap()
+    }
+}
+
+impl Mul<i64> for TokenAmount {
+    type Output = TokenAmount;
+
+    fn mul(self, rhs: i64) -> Self::Output {
+        if rhs < 0 {
+            // panic?
+        }
+
+        self * rhs as u64
+    }
+}
+
+impl Mul<u64> for &TokenAmount {
+    type Output = TokenAmount;
+
+    fn mul(self, rhs: u64) -> Self::Output {
+        *self * rhs
+    }
+}
+
+impl Mul<i64> for &TokenAmount {
```

Comment:
So, two options:

1. Don't implement mul, add, etc. and expect the user to call `checked_*` everywhere.
2. Implement them, but make the output something like `Result<TokenAmount, OverflowError>`. Users will need to write things like `(a * b)?`.
3. Implement them, but make the output some special `MaybeTokenAmount` with a `check(self) -> Result<TokenAmount, OverflowError>` function. Users would write, e.g., `(a * b + c).check()?`.

See https://crates.io/crates/checked for some inspiration.

Reply:
Okay, I went with 2 here. I think it works fine, except that I'm very liberally `unwrap`ing in places -- I'm not sure what the impact of that is for exit codes (though these are mostly fatal cases anyway).